x/ibc-rate-limit/rate_limit.go

```diff
@@ -0,0 +1,153 @@
+package ibcratelimit
+
+import (
+	"encoding/json"
+
+	errorsmod "cosmossdk.io/errors"
+	wasmkeeper "github.com/CosmWasm/wasmd/x/wasm/keeper"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	transfertypes "github.com/cosmos/ibc-go/v8/modules/apps/transfer/types"
+	clienttypes "github.com/cosmos/ibc-go/v8/modules/core/02-client/types"
+	"github.com/cosmos/ibc-go/v8/modules/core/exported"
+
+	"github.com/neutron-org/neutron/v4/x/ibc-rate-limit/types"
+)
+
+var (
+	msgSend = "send_packet"
+	msgRecv = "recv_packet"
+)
+
+func CheckAndUpdateRateLimits(ctx sdk.Context, contractKeeper *wasmkeeper.PermissionedKeeper,
+	msgType, contract string, packet exported.PacketI,
+) error {
+	contractAddr, err := sdk.AccAddressFromBech32(contract)
+	if err != nil {
+		return err
+	}
+
+	sendPacketMsg, err := BuildWasmExecMsg(
+		msgType,
+		packet,
+	)
+	if err != nil {
+		return err
```

Comment:
here and everywhere in this file, please wrap the errors with additional information about what failed (e.g. `fmt.Errorf(err, "failed to BuildWasmExecMsg")`)